pkg/build/api/validation/validation.go

```diff
@@ -141,23 +141,26 @@ func hasProxy(source *buildapi.GitBuildSource) bool {
 func validateSource(input *buildapi.BuildSource, isCustomStrategy bool) fielderrors.ValidationErrorList {
 	allErrs := fielderrors.ValidationErrorList{}
 
+	// Ensure that Git and Binary source types are mutually exclusive.
+	if input.Git != nil && input.Binary != nil && !isCustomStrategy {
+		allErrs = append(allErrs, fielderrors.NewFieldInvalid("git", "", "may not be set when binary is also set"))
+		allErrs = append(allErrs, fielderrors.NewFieldInvalid("binary", "", "may not be set when git is also set"))
+	}
```

Comment:
this was previously returning rather than continuing to validate, per @rhcarvalho 's suggestion that it wasn't really useful to pile more errors on the user at that point.